src/nxt_array.c

```diff
@@ -1,150 +1,253 @@
 
 /*
+ * Copyright (C) Evgenii Sokolov
  * Copyright (C) Igor Sysoev
  * Copyright (C) NGINX, Inc.
  */
 
 #include <nxt_main.h>
 
 
+static void nxt_array_elts_copy(void **dst_elts, nxt_array_t *src_array);
+static nxt_uint_t nxt_array_calc_alloc(nxt_uint_t nelts);
+static void *nxt_array_elts_realloc(nxt_array_t *array,
+                                    nxt_uint_t nalloc);
+
+
+/*
+ * nxt_array_create - Create a new array and return a pointer to it.
+ */
 nxt_array_t *
-nxt_array_create(nxt_mp_t *mp, nxt_uint_t n, size_t size)
+nxt_array_create(nxt_mp_t *mp, nxt_uint_t arr_size, size_t elt_size)
 {
-    nxt_array_t  *array;
+    nxt_uint_t     nalloc;
+    nxt_array_t    *array;
+    void           *elts;
 
-    array = nxt_mp_alloc(mp, sizeof(nxt_array_t) + n * size);
```

Comment:
The allocated memory is actually used until the first reallocation of memory. During the subsequent work of the program, the specified memory area was not used, remaining abandoned.

Reply:
You're trying to optimize a case that is rare in critical parts, while pessimizing the most common case.